transport/socket_transport.go

```diff
@@ -0,0 +1,503 @@
+	writeTimeout                                   = 5 * time.Second
+)
+
+var _ Server = (*SocketServer)(nil)
+
+/*
+SocketServer is a transport Server implementation that uses TCP sockets for communication between client and server.
+It can also be configured to use TLS.
+*/
+type SocketServer struct {
+	tlsConf             conf.TLSConfig
+	lock                sync.RWMutex
+	address             string
+	handlers            map[int]RequestHandler
+	started             bool
+	listener            net.Listener
+	acceptLoopExitGroup sync.WaitGroup
+	serverConnections   sync.Map
+}
+
+func NewSocketServer(address string, tlsConf conf.TLSConfig) *SocketServer {
+	return &SocketServer{
+		tlsConf:  tlsConf,
+		address:  address,
+		handlers: make(map[int]RequestHandler),
+	}
+}
+
+func (s *SocketServer) Start() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.started {
+		return nil
+	}
+	list, err := s.createNetworkListener()
+	if err != nil {
+		return err
+	}
+	s.listener = list
+	s.started = true
+	s.acceptLoopExitGroup.Add(1)
+	common.Go(s.acceptLoop)
+	log.Debugf("started socket transport server on address %s", s.address)
+	return nil
+}
+
+func (s *SocketServer) Stop() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if !s.started {
+		return nil
+	}
+	if s.listener != nil {
+		if err := s.listener.Close(); err != nil {
+			// Ignore
+		}
+	}
+	// Wait for accept loop to exit
+	s.acceptLoopExitGroup.Wait()
+	// Now close connections
+	s.serverConnections.Range(func(conn, _ interface{}) bool {
+		conn.(*serverConnection).stop()
+		return true
+	})
+	s.started = false
+	return nil
+}
+
+func (s *SocketServer) RegisterHandler(handlerID int, handler RequestHandler) bool {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	_, exists := s.handlers[handlerID]
+	if exists {
+		return false
+	}
+	s.handlers[handlerID] = handler
+	return true
+}
+
+func (s *SocketServer) getRequestHandler(handlerID int) (RequestHandler, bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	handler, exists := s.handlers[handlerID]
+	return handler, exists
+}
+
+func (s *SocketServer) Address() string {
+	return s.address
+}
+
+func (s *SocketServer) createNetworkListener() (net.Listener, error) {
+	var list net.Listener
+	var err error
+	var tlsConfig *tls.Config
+	if s.tlsConf.Enabled {
+		tlsConfig, err = conf.CreateServerTLSConfig(s.tlsConf)
+		if err != nil {
+			return nil, err
+		}
+		list, err = common.Listen("tcp", s.address)
+		if err == nil {
+			list = tls.NewListener(list, tlsConfig)
+		}
+	} else {
+		list, err = common.Listen("tcp", s.address)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return list, nil
+}
+
+func (s *SocketServer) acceptLoop() {
+	defer s.acceptLoopExitGroup.Done()
+	for {
+		conn, err := s.listener.Accept()
+		if err != nil {
+			// Ok - was closed
+			break
+		}
+		c := &serverConnection{
+			s:    s,
+			conn: conn,
+		}
+		s.serverConnections.Store(c, struct{}{})
+		c.start()
```

Comment:
Also log in trace? I've found connection issues the hardest to debug in Kafka esp in the context of cloud networking

Reply:
There is no trace level on the logger (info and debug warn and error)

I am a bit loathed to log too much as this would occur every time a connection is created